### PR TITLE
Christina & Danielle's - oo-ride-share

### PR DESCRIPTION
**Pull request by danimetz**
# OO Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | When  is the right time to raise an ArgumentError vs puts a "statement" instead. Talking about the implications of stopping the code.
Describe any examples of composition that you encountered in this project, if any |Had to decide if we wanted drivers to inherit from User. Also had to decide when setting up net_expenditures. How users, trip, and drivers all were tied together.
Describe the relationship between `User` and `Driver`|User is the Super class and Driver is the sub_class (driver inherits from User).
Describe a _nominal test_ that you wrote for this assignment. |the total_revenue test.
Describe an _edge case test_ that you wrote for this assignment | if all drivers are UNAVAILABLE in trip_request.
Describe a concept that you/your pair gained more clarity on as you worked on this assignment | TDD, OOD
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We discussed Patience and Pacing. Both were good as we were open about communicating ideas and when someone was stuck.

### CODE REVIEW

File: lib/trip.rb [hunk 1]
```diff
@@ -2,21 +2,35 @@
...
       @passenger = input[:passenger]
       @start_time = input[:start_time]
       @end_time = input[:end_time]
       @cost = input[:cost]
       @rating = input[:rating]
 
-      if @rating > 5 || @rating < 1
+
+      if input[:end_time] == nil
+        @end_time = input[:end_time]
```
Comment (dHelmgren) [inline lib/trip.rb:18]: You've already done the assignment, so this probably should have been an `if x && y` sort of a statement.

File: lib/trip.rb [hunk 2]
```diff
@@ -2,21 +2,35 @@
...
+      elsif input[:end_time] < input[:start_time]
+        raise ArgumentError.new("End time cannot be before Start time.")
```
Comment (dHelmgren) [inline lib/trip.rb:20]: Be careful doing this kind of checking after already making the assignment, as the person calling your code may not know that the assignment has already happened, and use the code as if you had prevented the assignment.

File: lib/user.rb [hunk 3]
```diff
@@ -9,12 +11,20 @@ def initialize(input)
...
     def add_trip(trip)
       @trips << trip
     end
+
+    def net_expenditures
+      trips.reduce(0){|sum, trip| sum + trip.cost }
+    end
+
+    def total_time_spent
+      trips.reduce{|sum, trip| sum.trip_duration + trip.trip_duration }
```
Comment (dHelmgren) [inline lib/user.rb:27]: I would have opted for the `sum` enumerable here instead. Also, what happens when a trip is still in progress?

File: lib/driver.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+      if [:AVAILABLE, :UNAVAILABLE].include? (input[:status])
+        @status = input[:status]
+      else
+        raise ArgumentError.new('That is not a valid status')
+      end
+    end
+
+    def add_driven_trip(driver_trip)
+      if driver_trip.class != RideShare::Trip
```
Comment (dHelmgren) [inline lib/driver.rb:29]: This is not a very Ruby way to ensure that you're getting what you want. Sandi Metz would be sad.

File: lib/driver.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    def total_revenue
+      revenue = driven_trips.reduce(0){|sum, trip| sum + (trip.cost - 1.65) }
+      revenue*= 0.8
+      return revenue.round(2)
+    end
+
+    def net_expenditures
+      money_spent = super
+      return money_spent - self.total_revenue
```
Comment (dHelmgren) [inline lib/driver.rb:52]: You can actually use the return value of a method instead of assigning it first, like you did with `self.total_revenue`. `super` will return a numeric value, so you can save space by writing `return super - self.total_revenue`.

File: specs/driver_spec.rb [hunk 6]
```diff
@@ -1,104 +1,138 @@
...
+            start_time: Time.parse("2016-08-08"),
+            end_time: Time.parse("2016-08-08"),rating: 1,cost: 5)
+            @driver.add_driven_trip(trip)
+            trip2 = RideShare::Trip.new(id: 8, driver: @driver, passenger: nil,
+              start_time: Time.parse("2016-08-08"),
+              end_time: Time.parse("2016-08-09"),
+              rating: 5,cost: 10)
+              @driver.add_driven_trip(trip2)
+            end
+            it "correctly calculates the total revenue that's positive" do
```
Comment (dHelmgren) [inline specs/driver_spec.rb:110]: A single test is not sufficient to make sure that this method is fully tested.

File: lib/trip_dispatcher.rb [hunk 7]
```diff
@@ -33,44 +38,130 @@ def load_users(filename)
...
 
-        trip = Trip.new(parsed_trip)
-        passenger.add_trip(trip)
-        trips << trip
-      end
+        in_progress_ride = Trip.new(parsed_trip)
 
-      return trips
-    end
+        new_driver.status = :UNAVAILABLE
```
Comment (dHelmgren) [inline lib/trip_dispatcher.rb:145]: This should probably be done as part of adding an in-progress trip to a `Driver` by the `Driver` instance itself.

File: lib/trip_dispatcher.rb [hunk 8]
```diff
@@ -33,44 +38,130 @@ def load_users(filename)
...
+
+        return available_drivers.first
+
+      end
+
+      def request_trip(user_id)
+        # The user ID will be supplied (this is the person requesting a trip)
+        @drivers.each do |driver|
+          if driver.id == user_id && driver.status == :AVAILABLE
+            driver.status = :UNAVAILABLE
```
Comment (dHelmgren) [inline lib/trip_dispatcher.rb:129]: This is not a good way to do this. It relies on messing with data that might not get put back if the method doesn't finish the way you expect. The better option is to make sure that `available_driver` can't give a driver whose id matches the `user_id`. What happens if you have an error in `find_passenger`? That driver's status is unavailable until someone can go in and fix it manually!

File: specs/trip_dispatch_spec.rb [hunk 9]
```diff
@@ -111,4 +107,60 @@
...
+      it "Add's an in-progress trip correctly driver, passenger & dispatched trips" do
+        trip =@dispatcher.request_trip(2)
+        expect(trip).must_be_instance_of RideShare::Trip
+        expect(trip.id).must_equal @dispatcher.trips.length
+        expect(trip.end_time).must_be_nil
+        expect(trip.rating).must_be_nil
+        expect(trip.cost).must_be_nil
+        expect(trip.driver).must_be_instance_of RideShare::Driver
+        expect(trip.passenger).must_be_kind_of RideShare::User
+        expect(trip.driver.status).must_equal :UNAVAILABLE
```
Comment (dHelmgren) [inline specs/trip_dispatch_spec.rb:125]: There are many more things you should be testing for the `request_trip` method:

- Does the trip reference the correct passenger and driver?
- Was the trip added to the trip lists of the driver and passenger?
- What if there is no available driver?

The last is an important edge case, and is exactly the sort of situation where software could do something unexpected. You might set up the test as follows:

```ruby
it 'raises an error if no driver is available' do
  dispatcher = RideShare::TripDispatcher.new
  dispatcher.drivers.each do |driver|
    # This may require an attr_writer in Driver
    driver.status = :UNAVAILABLE
  end
  proc {
    dispatcher.request_trip(2)
  }.must_raise StandardError
end
```

**Comment by dHelmgren** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Yes!
Answer comprehension questions | Your answer gets at this, but I want to call out this vocab explicitly. By composition we mean a _has-a_ or _has-many_ relationship. Examples in this project include "a TripDispatcher _has-many_ Drivers, Passengers and Trips", "a Driver _has-many_ Trips", "a Trip _has-a_ Driver".
**Wave 1** |
Appropriate use of Ruby's `Date`| Yep!
`Trip` has a helper method to calculate duration| Yep!
`User` (passenger) has a method to calculate total cost of all trips| Yes!
Tests for wave 1 | You don't seem to be testing all the methods that you were writing for this wave. I don't see tests for trip duration in particular.
**Wave 2** |
`Driver` inherits from `User` | yes!
`Driver` has `add_driven_trip` method | Yes! 
`Driver` has method to calculate average rating | Yes! 
`Driver` has method to calculate net expenditures and it uses `super` | Yes!
`Driver` has a method to calculate total revenue | Heck Yes!
Tests for wave 2 | Again, testing is worryingly insufficient. 
**Wave 3** |
`TripDispatcher` has a new method to create trips | Yes!
creating a trip in `TripDispatcher` relies on methods in `Driver` and `User` (passenger) to modify their own attributes | With the exception of driver availability.
Complex logic was correctly implemented | See my comments, you do something that works, but is VERY dangerous.
Tests for `request_trip` | See my comment. These aren't sufficient.
Methods from wave 1 and 2 handle incomplete trips | They do not.
Tests for wave 1 and 2 methods with incomplete trips | No tests to check.
**Wave 4 (Optional)** |
`TripDispatcher` now assigns trips to either the newest driver (no trips), or the driver who has not driven in the longest time | 
Appropriate helper methods were made to help with complex logic | 
Tests for wave 4 | 
**Overall** | I'm not sure what happened here. The code looks correct, but your tests are difficult to read due to poor code formatting, and many methods are barely tested if at all. We want you to engage in TDD, and part of that is writing tests that examine code from many points of potential failure, and the tests present here do not do that.
